File: cs/Problems/Day6.cs
namespace aoc24.Problems;

// https://adventofcode.com/2024/day/6
public sealed class Day6 : IProblem<int>
{
    public int Solve(string input) => CountGuardStepsOptimized(input);

    private readonly HashSet<Point> history = [];

    public int CountGuardStepsOptimized(ReadOnlySpan<char> input)
    {
        Span<char> inner = stackalloc char[Matrix<char>.SizeFor(input)];
        var matrix = Matrix<char>.CreateFrom(input, inner);
        history.Clear();

        var start = matrix.SeekItem('^').GetValueOrDefault();
        var current = new FacingPoint(Direction.Up, start);
        history.Add(start);

        while (true)
        {
            var next = current.MoveForward();
            var nextChar = matrix.ItemAt(next.Position);

            if (nextChar is null) // Went out of bounds.
                break;

            if (nextChar == '#')
            {
                current = TurnRight(current);
                continue;
            }

            current = next;
            history.Add(current.Position);
        }

        return history.Count;
    }

    private static FacingPoint TurnRight(FacingPoint facingPoint)
    {
        var newDirection = facingPoint.Direction switch
        {
            Direction.Up => Direction.Right,
            Direction.Down => Direction.Left,
            Direction.Right => Direction.Down,
            Direction.Left => Direction.Up,
            _ => facingPoint.Direction
        };
        return facingPoint with { Direction = newDirection };
    }
}

File: go/problems/day6.go
package problems

import (
	"image"
	"strings"

	"github.com/Meryovi/hobby-aoc-2024/go/shared"
)

type Day6 struct{}

var (
	up    = image.Pt(0, -1)
	down  = image.Pt(0, 1)
	right = image.Pt(1, 0)
	left  = image.Pt(-1, 0)
)

func (d Day6) Solve(input string) int {
	h := make(map[image.Point]image.Point)
	matrix := strings.Split(input, shared.NewLine)

	pt := seekStartPoint(matrix, '^')
	dir := up
	h[pt] = pt

	for {
		next := pt.Add(dir)
		if next.Y >= len(matrix) || next.X >= len(matrix[0]) || next.Y < 0 || next.X < 0 {
			break
		}

		c := matrix[next.Y][next.X]
		if c == '#' {
			dir = turnRight(dir)
			continue
		}

		pt = next
		h[pt] = pt
	}

	return len(h)
}

func seekStartPoint(matrix []string, char rune) image.Point {
	for y, l := range matrix {
		for x, c := range l {
			if c == char {
				return image.Pt(x, y)
			}
		}
	}
	return image.Pt(0, 0)
}

func turnRight(dir image.Point) image.Point {
	switch dir {
	case up:
		return right
	case down:
		return left
	case right:
		return down
	case left:
		return up
	default:
		return dir
	}
}
